lightning/src/ln/channelmanager.rs

```diff
@@ -906,20 +904,23 @@ impl<ChanSigner: ChannelKeys, M: Deref> ChannelManager<ChanSigner, M> where M::T
 		}
 
 		let mut chacha = ChaCha20::new(&rho, &[0u8; 8]);
+		let mut chacha_stream = ChaChaReader { chacha: &mut chacha, read: Cursor::new(&msg.onion_routing_packet.hop_data[..]) };
 		let (next_hop_data, next_hop_hmac) = {
-			let mut decoded = [0; 65];
-			chacha.process(&msg.onion_routing_packet.hop_data[0..65], &mut decoded);
-			let mut hmac = [0; 32];
-			hmac.copy_from_slice(&decoded[33..]);
-			match msgs::OnionHopData::read(&mut Cursor::new(&decoded[..33])) {
+			match msgs::OnionHopData::read(&mut chacha_stream) {
 				Err(err) => {
 					let error_code = match err {
 						msgs::DecodeError::UnknownVersion => 0x4000 | 1, // unknown realm byte
 						_ => 0x2000 | 2, // Should never happen
 					};
 					return_err!("Unable to decode our hop data", error_code, &[0;0]);
 				},
-				Ok(msg) => (msg, hmac)
+				Ok(msg) => {
+					let mut hmac = [0; 32];
+					if let Err(_) = chacha_stream.read_exact(&mut hmac[..]) {
+						return_err!("Unable to decode hop data", 0x4000 | 1, &[0;0]);
```

Comment:
IIRC (and this may be untrue, so double check), BADONION generally is used to imply "the previous hop peer from me is broken" not "the thing the original sender sent is nonsense", which this is, so we don't want to set BADONION. We could leave it as invalid_realm since we can't reasonably call this a TLV onion payload either, but I don't know that it matters which.

Reply:
"0x8000 (BADONION): unparsable onion encrypted by sending peer" (but don't say if sending peer is original sender or previous hop...).

You replaced it by PERM|22 on last tip ? I do think that BADONION|PERM|5 is the one which fits the best a) because it's the internal hmac so if wrong it's a failure of the original sender b) better for backward compatibility as this case can be triggered by old nodes too (and they shouldn't understand `invalid_onion_payload`). 

But if you disagree with BADONION meaning interpretation I'm fine like that.

Reply:
Hmm, maybe that's confusing, but certainly, at least for BADONION|5 and BADONION|6 its possible that the previous hop corrupted the onion before we got it (and we return differently - we pass the previous node a message to tell *them* to encrypt the failure onion).

In this case, this is *exactly* the thing that invalid_onion_payload should capture, I think - it says that we probably read the actual hop data in some bogus way and ended up without enough bytes left for the hmac.

Reply:
Hmmm but still this HMAC is encrypted to the actual processing node, so the previous node may have tweaked the packet and its HMAC but we won't decrypt in this case. I still think this error is only triggerable by the original sender.

So yes we agree on the error case, it's just BADONION is ambiguous..